docs/architecture/adr-022-custom-panic-handling.md

```diff
@@ -0,0 +1,207 @@
+externally and will allow developers use as many custom error handlers as they want. Implementation with tests
+can be found [here](https://github.com/cosmos/cosmos-sdk/pull/6053).
+
+#### Implementation details
+
+##### Recovery handler
+
+We add a `recover()` object handler type:
+
+```go
+type RecoveryHandler func(recoveryObj interface{}) error
+```
+
+`recoveryObj` is a return value of `recover()` function.
+Handler should type assert (or other methods) an object to define if object should be handled.
+`nil` should be returned if input object can't be handled by that `RecoveryHandler` (not a handler's target type).
+Not `nil` error should be returned if input object was handled and middleware chain execution should be stopped.
+
+An example:
+
+```go
+func exampleErrHandler(recoveryObj interface{}) error {
+    err, ok := recoveryObj.(error)
+    if !ok { return nil }
+    
+    if someSpecificError.Is(err) {
+        panic(customPanicMsg)
+    } else {
+        return nil
+    }
+}
+```
+
+This example breaks the application execution, but it also might enrich the error's context like the `OutOfGas` handler.
+
+##### Recovery middleware
+
+We also add a middleware type (decorator). That function type wraps `RecoveryHandler` and returns next middleware in
+execution chain and handler's `error`. Type is used to separate actual `recovery()` object handling from middleware
+chain processing.
+
+```go
+type recoveryMiddleware func(recoveryObj interface{}) (recoveryMiddleware, error)
+
+func newRecoveryMiddleware(handler RecoveryHandler, next recoveryMiddleware) recoveryMiddleware {
+    return func(recoveryObj interface{}) (recoveryMiddleware, error) {
+        if err := handler(recoveryObj); err != nil {
+            return nil, err
+        }
+        return next, nil
+    }
+}
+```
+
+Function receives a `recover()` object and returns:
+* (next `recoveryMiddleware`, `nil`) if object wasn't handled (not a target type) by `RecoveryHandler`;
+* (`nil`, not nil `error`) if input object was handled and other middlewares in the chain should not be executed;
+* (`nil`, `nil`) this is an invalid behaviour 'cause in that case panic recovery might not be properly handled;
+This can be avoided by always using a `default` as a rightmost middleware in chain (always returns an `error`'); 
+
+`OutOfGas` middleware example:
+```go
+func newOutOfGasRecoveryMiddleware(gasWanted uint64, ctx sdk.Context, next recoveryMiddleware) recoveryMiddleware {
+    handler := func(recoveryObj interface{}) error {
+        err, ok := recoveryObj.(sdk.ErrorOutOfGas)
+        if !ok { return nil }
+
+        return sdkerrors.Wrap(
+            sdkerrors.ErrOutOfGas, fmt.Sprintf(
+                "out of gas in location: %v; gasWanted: %d, gasUsed: %d", err.Descriptor, gasWanted, ctx.GasMeter().GasConsumed(),
+            ),
+        )
+    }
+    
+    return newRecoveryMiddleware(handler, next)
+}
+```
+
```

Comment:
I think adding `OutOfGas` and `Default` examples here out of the blue make this ADR hard to read and a bit confusing. I'd avoid them or change how you want to show the concrete changes to baseapp.

Reply:
I think the "BaseApp changes" section should be preserved as it shows how exactly the whole idea works: creating a `default` middleware and embedding it to the panic recovery section (`runTx` method).
Regarding middleware examples. I'm not sure about removing them too, as those functions are used in examples ("BaseApp changes" for instance). But I do believe you're more experienced in ADR design and I can remove those if needed.